### PR TITLE
Add German Translation

### PR DESCRIPTION
**Pull request by UnlimitedCookies**
Howdy!

This PR adds german translation.

I still have one question:
In `No need to include URL, just everything after. Args are delimited using two commas like so: ,,` you probably mean the FQDN, because the URL is the entire thing while the FQDN is only the part before the file path slash (`example.com` /notFQDN). I have changed this accordingly in my translation.

Best regards

### CODE REVIEW

**Comment by Tzahi12345** [comment]
I should have been more clear with that hint, basically what I mean is that you shouldn't include the video URL in the custom args. It's not even a necessary comment as intuitively custom args wouldn't include the video URL. Oh well!

Let me know when it's ready to merge and I'll do so!

File: src/assets/i18n/messages.de.json [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+  "d3f02f845e62cebd75fde451ab8479d2a8ad784d": "Ansehen",
+  "4a9889d36910edc8323d7bab60858ab3da6d91df": "Nur Audio",
+  "96a01fafe135afc58b0f8071a4ab00234495ce18": "Multi-Download Modus",
+  "6a21ba5fb0ac804a525bf9ab168038c3ee88e661": "Download",
+  "6a3777f913cf3f288664f0632b9f24794fdcc24e": "Abbrechen",
+  "322ed150e02666fe2259c5b4614eac7066f4ffa0": "Erweitert",
+  "b7ffe7c6586d6f3f18a9246806a7c7d5538ab43e": "Simulierter Befehl:",
+  "4e4c721129466be9c3862294dc40241b64045998": "Benutzerdefinierte Argumente verwenden",
+  "ad2f8ac8b7de7945b80c8e424484da94e597125f": "Benutzerdefinierte Argumente",
+  "a6911c2157f1b775284bbe9654ce5eb30cf45d7f": "Der FQDN muss nicht angegeben werden, sondern nur der Teil danach. Argumente werden wie folgt getrennt: ,,",
```
Comment (SuperSandro2000) [inline src/assets/i18n/messages.de.json:28]:
```suggestion
  "a6911c2157f1b775284bbe9654ce5eb30cf45d7f": "Der FQDN muss nicht angegeben werden, sondern nur der Teil danach. Argumente werden mit zwei Kommata getrennt: ,,",
```

File: src/assets/i18n/messages.de.json [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+  "3a92a3443c65a52f37ca7efb8f453b35dbefbf29": "Benutzerdefinierte Ausgabe verwenden",
+  "d9c02face477f2f9cdaae318ccee5f89856851fb": "Benutzerdefinierte Ausgabe",
+  "fcfd4675b4c90f08d18d3abede9a9a4dff4cfdc7": "Dokumentation",
+  "19d1ae64d94d28a29b2c57ae8671aace906b5401": "Der Pfad ist relativ zum Konfigurations-Download-Pfad. Dateiendung nicht angeben.",
```
Comment (SuperSandro2000) [inline src/assets/i18n/messages.de.json:32]:
```suggestion
  "19d1ae64d94d28a29b2c57ae8671aace906b5401": "Der Pfad ist relativ zum Konfigurations-Download-Pfad. Dateiendung nicht mit angeben.",
```
Not sure about this one.

File: src/assets/i18n/messages.de.json [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+  "a67e7d843cef735c79d5ef1c8ba4af3e758912bb": "Hochgeladen am:",
+  "f4e529ae5ffd73001d1ff4bbdeeb0a72e342e5c8": "Schließen",
+  "ca3dbbc7f3e011bffe32a10a3ea45cc84f30ecf1": "ID:",
+  "e684046d73bcee88e82f7ff01e2852789a05fc32": "Anzahl:",
+  "321e4419a943044e674beb55b8039f42a9761ca5": "Info",
+  "826b25211922a1b46436589233cb6f1a163d89b7": "Löschen",
+  "34504b488c24c27e68089be549f0eeae6ebaf30b": "Löschen und zur Blacklist hinzufügen",
+  "121cc5391cd2a5115bc2b3160379ee5b36cd7716": "Einstellungen",
+  "801b98c6f02fe3b32f6afa3ee854c99ed83474e6": "URL",
+  "54c512cca1923ab72faf1a0bd98d3d172469629a": "URL, über die auf diese Applikation zugegriffen wird, ohne den Port.",
```
Comment (SuperSandro2000) [inline src/assets/i18n/messages.de.json:57]:
```suggestion
  "54c512cca1923ab72faf1a0bd98d3d172469629a": "URL, über die auf diese Applikation zugegriffen wird, ohne Port.",
```

File: src/assets/i18n/messages.de.json [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+  "cb2741a46e3560f6bc6dfd99d385e86b08b26d72": "Port",
+  "22e8f1d0423a3b784fe40fab187b92c06541b577": "Der gewünschte Port. Standard ist 17442.",
+  "d4477669a560750d2064051a510ef4d7679e2f3e": "Multi-user",
```
Comment (SuperSandro2000) [inline src/assets/i18n/messages.de.json:60]:
```suggestion
  "d4477669a560750d2064051a510ef4d7679e2f3e": "Multi-User",
```

File: src/assets/i18n/messages.de.json [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+  "6b995e7130b4d667eaab6c5f61b362ace486d26d": "Globale benutzerdefinierte Argumente für Downloads auf der Startseite. Argumente werden durch zwei Kommas voneinander getrennt: ,,",
+  "0ba25ad86a240576c4f20a2fada4722ebba77b1e": "Downloader",
+  "61f8fd90b5f8cb20c70371feb2ee5e1fac5a9095": "Titel der Kopfzeile",
+  "78d3531417c0d4ba4c90f0d4ae741edc261ec8df": "Dateimanager aktivieren",
+  "a5a1be0a5df07de9eec57f5d2a86ed0204b2e75a": "Download-Manager aktivieren",
+  "c33bd5392b39dbed36b8e5a1145163a15d45835f": "Qualitätsauswahl erlauben",
+  "bda5508e24e0d77debb28bcd9194d8fefb1cfb92": "Nur Download Modus",
+  "09d31c803a7252658694e1e3176b97f5655a3fe3": "Multi-Download Modus erlauben",
+  "d8b47221b5af9e9e4cd5cb434d76fc0c91611409": "Einstellungen durch Pin schützen",
+  "f5ec7b2cdf87d41154f4fcbc86e856314409dcb9": "Neuen PIN festlegen",
```
Comment (SuperSandro2000) [inline src/assets/i18n/messages.de.json:94]:
```suggestion
  "f5ec7b2cdf87d41154f4fcbc86e856314409dcb9": "Neuen Pin festlegen",
```

Comment (GlassedSilver) [reply src/assets/i18n/messages.de.json:94]: Pin statt PIN? Es ist ein Akronym. Von daher... Eigentlich [PIN](https://www.duden.de/rechtschreibung/PIN)

Comment (SuperSandro2000) [reply src/assets/i18n/messages.de.json:94]: Dann sollte Pin in Zeile 93 auch PIN sein.

File: src/assets/i18n/messages.de.json (hunk 2 above)
Comment (UnlimitedCookies) [reply src/assets/i18n/messages.de.json:32]: This is not necessary imo. Let me know when you think differently.

File: src/assets/i18n/messages.de.json (hunk 4 above)
Comment (UnlimitedCookies) [reply src/assets/i18n/messages.de.json:60]: Multi-Benutzer klingt ziemlich doof 😋

File: src/assets/i18n/messages.de.json (hunk 2 above)
Comment (GlassedSilver) [reply src/assets/i18n/messages.de.json:32]: "Dateiendung auslassen" wäre auch eine Möglichkeit? :)

File: src/assets/i18n/messages.de.json (hunk 4 above)
Comment (UnlimitedCookies) [reply src/assets/i18n/messages.de.json:60]: Ich hab noch Modus angehängt. Den im original heißt es `Multi-user mode`

File: src/assets/i18n/messages.de.json (hunk 5 above)
Comment (UnlimitedCookies) [reply src/assets/i18n/messages.de.json:94]: Beide sind jetzt PIN.

File: src/assets/i18n/messages.de.json (hunk 4 above)
Comment (GlassedSilver) [reply src/assets/i18n/messages.de.json:60]: Multiuser wäre glaube ich die "korrekte" deutsche Schreibweise, allerdings bin auch ich für die Trennung mit Bindestrich für bessere Lesbarkeit. So ist das Wort auch sonst in deutschen IT-Publikationen anzutreffen.

Multi-Benutzer klingt ein bisschen nach "Rechenmaschine", wenn ihr wisst was ich meine. :P

Comment (UnlimitedCookies) [reply src/assets/i18n/messages.de.json]: Jo, seid ihr also mit `Multi-User Modus` einverstanden?

**Comment by UnlimitedCookies** [comment]
> I should have been more clear with that hint, basically what I mean is that you shouldn't include the video URL in the custom args. It's not even a necessary comment as intuitively custom args wouldn't include the video URL. Oh well!
> 
> Let me know when it's ready to merge and I'll do so!

1. Hmm, okay. @Tzahi12345 Can you clarify the English original, so that I can translate it properly into German? :)

2. And btw - I found it a little annoying that we have to maintain two files that essentially do the same (the xliff and json). That's because every change someone requested I had to change in both files - running your converter script doesn't make it much quicker. Maybe someone has an Idea on how to improve this.

Thanks!

File: src/assets/i18n/messages.de.json (hunk 5 above)
Comment (UnlimitedCookies) [reply src/assets/i18n/messages.de.json:94]: So viele Deutsche hier, doch die Deutschen brauchen eine Übersetzung eigentlich am wenigsten :D

**Comment by Tzahi12345** [comment]
>Hmm, okay. @Tzahi12345 Can you clarify the English original, so that I can translate it properly into German? :)

Will do later today when I get the chance! You can actually jump the gun on this one and translate this string into German:

`Args are delimited using two commas like so: ,,`

I think any more info in that hint would confuse users, so I think we're good with that. I'll update the original tonight to be the above string tonight!

>And btw - I found it a little annoying that we have to maintain two files that essentially do the same (the xliff and json). That's because every change someone requested I had to change in both files - running your converter script doesn't make it much quicker. Maybe someone has an Idea on how to improve this.

It is annoying, I decided to go with that route for a few reasons:

1. JSONs are significantly smaller than XLIFFs (not that big of a deal, XLIFFs are just a few hundred KB)
2. To load languages at runtime (meaning without having to rebuild the app), you need a JSON format. If you open [this article](https://blog.ninja-squad.com/2019/12/10/angular-localize/) and scroll down to `Runtime translations`, you'll see how it's generally done. You can see where we do this [in main.ts](https://github.com/Tzahi12345/YoutubeDL-Material/blob/master/src/main.ts) by calling Angular's loadTranslations() function. So JSON's just play nicer with the API.
3. Why even keep the XLIFF? Well, the JSON doesn't have the original string, it just contains the hash of the ID and that's it. We want to keep the XLIFF as a reference in case any strings do get updated or the translated one needs a second look.

The script is really there to just help convert the XLIFF to a JSON. Without the script, the initial JSON generation would take forever as you'd need to build it yourself (ew).

So that's basically the system we have now. It's not a pretty process, and like you, I'm super open to any suggestions on how to make this better (maybe generate the JSON on the fly? I wrote the code for that already anyways...)

----

Anyways, let me know when you want this merged! Very excited to get German support :)

**Comment by UnlimitedCookies** [comment]
Okay... I have now translated it 1:1, since it's not the job of the translation to make such changes.

I totally understand why you have gone this route, I was just looking for an easier way, so more people might add a translation :)

Other than that, I think we're ready for a merge. (If @GlassedSilver and @SuperSandro2000 don't have anything to add).

File: src/assets/i18n/messages.de.json (hunk 5 above)
Comment (GlassedSilver) [reply src/assets/i18n/messages.de.json:94]: Ich hab schon gesagt, dass ich das Projekt auch übersetzen würde, aber es vermutlich kaum anrühren würde nach dem ersten Commit, weil ich Software wie diese in Englisch nutze. XD

YT selbst ist auch EN, sowie 90% des Contents den ich schaue. der Rest ist eher JP als DE. :,D

Aber ist doch eigentlich super umso mehr die Software nutzen können, desto besser. Füllt definitiv ne Lücke auf dem Softwaremarkt. :)

**Comment by GlassedSilver** [comment]
I only skimmed so far, but seems alright, so giving this a rocket for lift off. :D

File: src/assets/i18n/messages.de.json (hunk 5 above)
Comment (UnlimitedCookies) [reply src/assets/i18n/messages.de.json:94]: Da hast du recht, ich werde die Software auch weiterhin in Englisch nutzen. Die Deutschen sind durch ihre gute englische Ausbildung aber meist nicht auf Übersetzungen angewiesen.

**Comment by Tzahi12345** [comment]
With @GlassedSilver's official rocket, I think this is ready to merge. Thank you so much for this contribution! The more translations there are, the more welcoming this project is. There seems to be quite a few users from Germany in particular, so extra points for that.

Feel free to submit more PRs if more text is added and the translation gets out of date (basically there will just be untranslated text). If you want, I can make you the official maintainer for the German translations, up to you (I'll put a spot in the README and I'll own English/Spanish).

Thanks again @UnlimitedCookies and everyone else in this thread for the hard work!

**Comment by UnlimitedCookies** [comment]
> Feel free to submit more PRs if more text is added and the translation gets out of date (basically there will just be untranslated text). If you want, I can make you the official maintainer for the German translations, up to you (I'll put a spot in the README and I'll own English/Spanish).

Sure.

**Comment by UnlimitedCookies** [comment]
One more question:
When you make a breaking change to the meaning of the original string, do you assign a new ID for translation, so I see that this has to be retranslated?

Like in #129 , where the meaning has been significantly changed.

**Comment by Tzahi12345** [comment]
The bottom of [this section](https://angular.io/guide/i18n#translate-each-translation-file) in the Angular docs explains it well:

>Don't change the IDs for translation units. Each `id` is generated by Angular and depends on the content of the template text and its assigned meaning. If you change either the text or the meaning, then the `id` changes. For more about managing text updates and IDs, see the previous section on [custom IDs.](https://angular.io/guide/i18n#custom-id)

So basically, once the text changes, the ID will change too. Thankfully, this doesn't happen very often, but if it does, it will leave that string of text untranslated. I now realize that if I update the hint as I suggested above, it would break your translation (so will #129 once that is compiled).

We can fix this by either assigning our own custom tags for each string in the app (tedious), or being more careful/coordinated about text updates. Also, it's not the end of the world if the translation is missing a couple strings out of hundreds for a short period of time. So no pressure on this front. Thanks for taking up the maintainer position! If you ever feel it's too much or you want to hand it off to someone else, let me know :)

I'll make sure to avoid frivolous changes to strings to avoid giving us unnecessary work in the future. Let me know if you have any other questions.

**Comment by UnlimitedCookies** [comment]
To streamline the translation process the project could use a central Translation management system (TMS).
Popular ones are [crowdin](https://crowdin.com/page/open-source-project-setup-request) and [lokalize](https://lokalise.com/) I think they offer free plans for open source projects.

They're unfortunately not open source (but Github isn't OSS either!). These are open source alternatives, but some require to be selfhosted.
[Mozilla Pontoon](https://github.com/mozilla/pontoon/)
[Pootle](https://pootle.translatehouse.org/discover.html)

Check out [this article](https://opensource.com/article/17/6/open-source-localization-tools) for more information. :)

Lmk what you guys think.

**Comment by Tzahi12345** [comment]
Just checked out both crowdin and lokalize, I like both but leaning towards crowdin (they suggest translations which is nice), but I'll play around with both.

But I also really like the idea of self-hosting one (Pontoon looks cool), issue is where will it be hosted? The best long-term approach might be to get the free OSS plan for one of these commercial products, but what do I know. Anyways, those are my thoughts. I already asked crowdin for one of those licenses, so we'll see where that goes.

**Comment by UnlimitedCookies** [comment]
Hey, has there been any update on this?

You could also consider https://weblate.org/en/ , who offer free hosting (for libre projects) and are OSS.

**Comment by Tzahi12345** [comment]
I actually signed up for crowdin and lokalize, but neither got back to me on the free service for OSS, not sure why. But I'll check out weblate, I couldn't find the free hosting option so lmk where that is! I do still think this needs to be done, I'll see if I can find time this week to get it up and running.

**Comment by UnlimitedCookies** [comment]
Sounds good.

If you scroll down further on their homepage you can find an option saying "Gratis Weblate for libre projects"
See in the picture below.
image
image

**Comment by Tzahi12345** [comment]
Not sure how I missed that! I went ahead and set up the project, you should be good to go to "suggest" translations for untranslated strings, right now my ability to give access is limited so hopefully once they approve the project I can give permissions.

https://hosted.weblate.org/projects/youtubedl-material/ytdl-material/

If it works well on your end, then I'll happily add it to the docs. I need to fix the Spanish translation since it appears it didn't get imported for whatever reason. Worst case I'll re-do the translations :P

**Comment by UnlimitedCookies** [comment]
Looking awesome, I've now suggested the missing translations.
And I can also already see chinese, which is awesome (though I am wondering why the file name is `src/assets/i18n/messages.zh-Hans.xlf`).

Weird that it didn't work for spanish.

**Comment by Tzahi12345** [comment]
>And I can also already see chinese, which is awesome (though I am wondering why the file name is src/assets/i18n/messages.zh-Hans.xlf).

I'll try to correct that. But don't give me credit for it! It was translations I copied off of @TyRoyal. I also added French which @Halfayd graciously provided (#204). 

I got the Spanish translations working again, turns out the erroneous XML elements that it had messed it up for whatever reason. After fixing it and uploading it, it was fixed. But during that process I cleared the local repo that Weblate had, and I saw that the Russian translation that was added got deleted. I hope that I didn't delete any of your progress there, I'm really sorry if I did.

Thank you though for showing me this great service, they already approved the project so we're good to go with them :)

**Comment by TyRoyal** [comment]
> > And I can also already see chinese, which is awesome (though I am wondering why the file name is src/assets/i18n/messages.zh-Hans.xlf).
> 
> I'll try to correct that. But don't give me credit for it! It was translations I copied off of @TyRoyal. I also added French which @Halfayd graciously provided (#204).
> 
> I got the Spanish translations working again, turns out the erroneous XML elements that it had messed it up for whatever reason. After fixing it and uploading it, it was fixed. But during that process I cleared the local repo that Weblate had, and I saw that the Russian translation that was added got deleted. I hope that I didn't delete any of your progress there, I'm really sorry if I did.
> 
> Thank you though for showing me this great service, they already approved the project so we're good to go with them :)

Wow, `messages.zh-Hans.xlf` could renamed as `messages.zh.xlf` because there are 2 big versions `Chinese (Simplified)` zh-Hans and `Chinese (Traditional)` zh-Hant.

Also I found that `Custom playlists` component is not translatable in `messages.en.xlf`

[https://github.com/Tzahi12345/YoutubeDL-Material/tree/master/src/app/components/custom-playlists](url)

**Comment by UnlimitedCookies** [comment]
Magnificent. Good that we finally found something that's working.

@Tzahi12345 can you perhaps approve me to make changes to the German
translation? Or accept my changes? ;)

@TyRoyal I'm sorry, I don't get what you're saying. I was just pointing out
that it was not in compliance with the two-letter codes for languages, or
is there something different for chinese?

**Comment by Tzahi12345** [comment]
@UnlimitedCookies done! translate away :)

@TyRoyal Fixed issue with custom playlists! [Here's the commit](https://github.com/Tzahi12345/YoutubeDL-Material/commit/7dcc38c26d12f0240e1593ba20899fea2b98c8ba). Also, if you want to be the official translator for Chinese, I'll give you extra permissions in Weblate. Either way, you can [update/contribute your translations here](https://hosted.weblate.org/translate/youtubedl-material/ytdl-material/zh_Hans/?q=state:empty). Awesome work, thanks for helping us get Chinese language support!

**Comment by TyRoyal** [comment]
@UnlimitedCookies 
Yes, Chinese (Simplified) and Chinese (Traditional) are different.

For example,
`简体中文` is Simplified Chinese as code zh-Hans
`簡體中文` is Traditional Chinese as code zh-Hant
They are different and the code is `RFC 4646`

But in `ISO 639-1` , which this program has been using, `zh-Hans` and `zh-Hant` have the same name `zh`

@Tzahi12345 
Thx! It's my pleasure to be the official translator for Chinese~
But I would like to use another account to be the one of the contributors (which might be helpful for applying for MSc) 😁

**Comment by Tzahi12345** [comment]
@TyRoyal Awesome! Let me know what your email is, or [register here](https://hosted.weblate.org/) and I'll add your username to the project

**Pull request merged**